### ISSUE TITLE
jQuery Passthrough with jQuery Isotope plugin

### ISSUE THREAD

**Issue by jzbruno** [angular-ui/angular-ui-OLDREPO#319]
I am trying to get jQuery Passthrough to use a jQuery plugin called Isotope which orders and arranges elements. http://isotope.metafizzy.co/ but I am not having any luck.

Here is a basic example without Angular, it just orders the items in a row alphabetically
http://jsfiddle.net/jzbruno/SZZTh/9/

But when I try to use this with AngularJS through Angular-UI the data is not rendered. (you can see the data rendered by removing the ui-jq directive.
http://jsfiddle.net/jzbruno/SZZTh/8/

Any ideas? Anything obvious stand out? Thanks.

**Comment by ProLoser** [comment]
`ui-jq` will not work for this as the children elements have not been rendered by the time the container has been rendered (rendering is done from TOP to BOTTOM, PARENT to CHILDREN).

You may need to create your own directive or suggest ways in which to expand the versatility of `ui-jq` (I've been tossing around the idea of having a 'delay' option that when enabled, sort of hackishly bypasses most rendering bugs.

Also worth reading: https://groups.google.com/forum/?fromgroups=#!topic/angular/w3g7YMuv34E

**Comment by abanctelchevrel** [comment]
Any use of directives (in your case ngRepeat) within the jQuery passthrough will mostly fail, due to how Angular renders the markup.
I had a similar issue trying to use the [tagit jquery plugin](http://webspirited.com/tagit/) via angular-ui jQuery passthrough.

A dirty way of solving your issue would be to wrap the call to the plugin in a setTimeout(0), replacing [this line](https://github.com/angular-ui/angular-ui/blob/master/modules/directives/jq/jq.js#L45)

``` javascript
        elm[attrs.uiJq].apply(elm, linkOptions);
```

with

``` javascript
        setTimeout(function(){
          elm[attrs.uiJq].apply(elm, linkOptions);
        }, 0);
```

**Comment by cmawhorter** [comment]
Another possible solution is to wrap the jquery function:

``` javascript
var _$_fn_dataTable = $.fn.dataTable;
$.fn.dataTable = function() {
  var self = this
    , args = arguments;
  setTimeout(function() {
    _$_fn_dataTable.apply(self, arguments);
  }, 0);
};
```

I think it'd be nice to have something like the following so that we can get this when we need it:

``` html
<table ui-jq-delay="dataTable">
  <tr ng-repeat="row in rows">
```

Just started using angular, so I might be way off here.

**Comment by ProLoser** [comment]
Actually I have this in a PR waiting to be merged soon.

You just gave me a different thought though. I was going to do `ui-jq="plugin" ui-defer` but maybe we should do it your way with `ui-jq-defer="plugin"` instead?

**Comment by cmawhorter** [comment]
Personally, I like ui-jq-defer just so there is no question about the intent.

**Comment by surjikal** [comment]
Would `setTimeout` be called only once? If so, I'm not sure if the solution you guys are proposing would cover my particular use case. Basically, I've got a unordered list that is populated by selecting a model from another list. So here, I think you would need to call `setTimeout` when the selected model is changed, right?

**Comment by ProLoser** [comment]
You are talking about 2 different issues.  

The PR contains ui-defer and ui-refresh. The refresh takes an expression to watch and re-initialize the plugin on (this may not help everyone). The defer simply says that in order to initialize, you must do so in a setTimeout.

I prefer ui-jq-defer except for 1 problem: I'd have to make a new directive with all the same code.
Now this isn't impossible, it's just unsexy. Granted, i've also done it before and can still make it DRY, I just don't like that I have to.

A secondary attribute can be used inside the original directive.

Also, if it's a secondary property, you can define the option in the global config. This means you don't have to keep remembering if this plugin needs to be deferred every time you attach it.  

## 

Dean Sofer
DeanSofer.com
714.900.2254

**Comment by surjikal** [comment]
Can you link the PR here, please? I'd like to take a look at it! Sounds like the combination of `ui-defer` and `ui-refresh` is what I need.

EDIT: Nevermind, found it! https://github.com/angular-ui/angular-ui/pull/347

**Comment by ProLoser** [comment]
If you guys want to help me FINALLY merge it in please try to figure out why some of the unit tests won't pass.

**Comment by ProLoser** [comment]
Merged! Next release will have this in the lib. Or you can pull the latest and build locally!

**Comment by cmawhorter** [comment]
Glad to see that!  And btw; re. unsexiness.  You could just create:

```
MyObject.prototype.someFunc = function() { };
MyObject.prototype.someFuncDeferred = function() { 
  var self = this, args = arguments;
  setTimeout(function() { MyObject.prototype.someFunc.apply(this, args); }, 0);
};
```

<3 js

**Comment by ProLoser** [comment]
That's an interesting workaround @cmawhorter. I will keep it in mind for the future.

**Comment by scottromack** [comment]
Please post a working example when available. angular + isotope would be very useful!

**Comment by ProLoser** [comment]
I started to put a demo together and found a bug I thought I had fixed. That's what I got for not having comprehensive unit tests.

**Comment by surjikal** [comment]
You're fired.

**Comment by ProLoser** [comment]
Seriously. Now I'm stuck trying to get the damn unit tests working again

**Comment by ProLoser** [comment]
#379
